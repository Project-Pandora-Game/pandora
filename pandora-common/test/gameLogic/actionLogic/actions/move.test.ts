import { describe, expect, it } from '@jest/globals';
import { AssertNotNullable } from '../../../../src/index.ts';
import { TestAssetsLoadAssetManager } from '../../../assets/_testData/testAssetsDefinition.ts';
import { TestCreateCharacterState, TestCreateGlobalState } from '../../../assets/assetsTestingHelpers.ts';
import { TestCreateGameLogicCharacter } from '../../character/characterTestingHelpers.ts';
import { TestActionExpectFail, TestActionExpectValidResult, TestCreateActionContext, TestDoImmediateAction, TestStateExtractAssets } from '../actionTestingHelpers.ts';

describe('ActionMoveItem', () => {
	it('Reorders item downwards', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/panties/style1',
				'a/panties/style1',
				'a/headwear/top_hat',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: -1,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Reorders item downwards by more than one', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/panties/style1',
				'a/panties/style1',
				'a/headwear/top_hat',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: -2,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Reorders item upwards', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/headwear/top_hat',
				'a/panties/style1',
				'a/panties/style1',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: 1,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Reorders item upwards by more than one', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/headwear/top_hat',
				'a/panties/style1',
				'a/panties/style1',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: 2,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Fails if reordering between bodyparts', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/panties/style1',
				'a/panties/style1',
				'a/headwear/top_hat',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: -4,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});

	it('Fails if reordering out of bounds downwards', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/panties/style1',
				'a/panties/style1',
				'a/headwear/top_hat',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: -1000,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});

	it('Fails if reordering out of bounds upwards', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, null, [
			(room) => TestCreateCharacterState(assetManager, character, room, [
				'a/headwear/top_hat',
				'a/panties/style1',
				'a/panties/style1',
			]),
		]);

		const topHatId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(topHatId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'move',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: topHatId,
				},
				shift: 1000,
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});
});
