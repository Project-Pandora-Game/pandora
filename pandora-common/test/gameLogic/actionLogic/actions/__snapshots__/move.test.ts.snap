// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMoveItem Fails if reordering between bodyparts 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;

exports[`ActionMoveItem Fails if reordering between bodyparts: problems 1`] = `
[
  {
    "result": "validationError",
    "validationError": {
      "problem": "bodypartError",
      "problemDetail": "incorrectOrder",
    },
  },
]
`;

exports[`ActionMoveItem Fails if reordering out of bounds downwards 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;

exports[`ActionMoveItem Fails if reordering out of bounds downwards: problems 1`] = `
[
  {
    "reason": undefined,
    "result": "invalidAction",
  },
]
`;

exports[`ActionMoveItem Fails if reordering out of bounds upwards 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Fails if reordering out of bounds upwards: problems 1`] = `
[
  {
    "reason": undefined,
    "result": "invalidAction",
  },
]
`;

exports[`ActionMoveItem Reorders item downwards 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;

exports[`ActionMoveItem Reorders item downwards 2`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Reorders item downwards by more than one 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;

exports[`ActionMoveItem Reorders item downwards by more than one 2`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Reorders item upwards 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Reorders item upwards 2`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Reorders item upwards by more than one 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionMoveItem Reorders item upwards by more than one 2`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;
